s3prl/downstream/a2o-vc-vcc2020/expert.py

```diff
@@ -0,0 +1,233 @@
+            records["wav_paths"] += wav_paths
+            records["wavs"] += wavs
+        else:
+            predicted_features, predicted_feature_lengths = self.model(input_features, input_feature_lengths, acoustic_features_padded.to(device))
+
+        # loss calculation (masking and normalization are done inside)
+        loss = self.objective(predicted_features,
+                              acoustic_features_padded,
+                              predicted_feature_lengths,
+                              acoustic_feature_lengths,
+                              device)
+        records['loss'].append(loss.item())
+
+        return loss
+
+
+    # interface
+    def log_records(self, split, records, logger, global_step, batch_ids, total_batch_num, **kwargs):
+        loss = torch.FloatTensor(records['loss']).mean().item()
+
+        if split in ["dev", "test"]:
+
+            hdf5_save_dir = os.path.join(self.expdir, str(global_step), split, "hdf5")
+            wav_save_dir = os.path.join(self.expdir, str(global_step), split, "wav")
+            os.makedirs(hdf5_save_dir, exist_ok=True)
+            os.makedirs(wav_save_dir, exist_ok=True)
+
+            for i, wav_path in enumerate(records["wav_paths"]):
+                length = int(records["feature_lengths"][i])
+                fbank = np.array(records["predicted_features"][i])[:length]
+
+                # save generated features into hdf5 files
+                hdf5_save_path = os.path.join(hdf5_save_dir, "_".join(wav_path.split("/")[-2:]).replace(".wav", ".h5"))
+                write_hdf5(hdf5_save_path, "feats", fbank)
+
+                # mel fbank -> linear spectrogram
+                spc = logmelspc_to_linearspc(
+                    fbank,
+                    fs=self.datarc["fbank_config"]["fs"],
+                    n_mels=self.datarc["fbank_config"]["n_mels"],
+                    n_fft=self.datarc["fbank_config"]["n_fft"],
+                    fmin=self.datarc["fbank_config"]["fmin"],
+                    fmax=self.datarc["fbank_config"]["fmax"],
+                )
+                # apply griffin lim algorithm
+                y = griffin_lim(
+                    spc,
+                    n_fft=self.datarc["fbank_config"]["n_fft"],
+                    n_shift=self.datarc["fbank_config"]["n_shift"],
+                    win_length=self.datarc["fbank_config"]["win_length"],
+                    window=self.datarc["fbank_config"]["window"],
+                    n_iters=self.datarc["fbank_config"]["gl_iters"],
+                )
+                # save generated waveform
+                wav_save_path = os.path.join(wav_save_dir, "_".join(wav_path.split("/")[-2:]))
+                write(
+                    wav_save_path,
+                    self.datarc["fbank_config"]["fs"],
+                    (y * np.iinfo(np.int16).max).astype(np.int16),
+                )
+
+        print(f'{split} loss: {loss:.6f}')
+
+        save_names = []
```

Comment:
Just confirming, so we will not save any checkpoints since you already saved all the inferenced files into expdir right?

Reply:
We actually are saving the checkpoints, however not used in the evaluation.

Reply:
OK I find it. You save the latest 10 checkpoints.